File: core/core-application/src/test/java/com/nmh/commerce/coupon/CouponStockManagerTest.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.coupon.mock.FakeCouponStockRepository;
import com.nmh.commerce.domain.Quantity;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import static org.assertj.core.api.BDDAssertions.then;

class CouponStockManagerTest {

    private CouponStockManager couponStockManager;
    private CouponStockRepository stockRepository;

    @BeforeEach
    void setUp() {
        FakeCouponStockRepository fakeCouponStockRepository = new FakeCouponStockRepository();
        this.stockRepository = fakeCouponStockRepository;
        this.couponStockManager = CouponStockManager.builder()
            .stockRepository(fakeCouponStockRepository)
            .build();
        CouponStock stock1 = CouponStock.builder()
            .id(1L)
            .couponId(1L)
            .remainingQuantity(Quantity.of(1))
            .build();
        CouponStock stock2 = CouponStock.builder()
            .id(2L)
            .couponId(2L)
            .remainingQuantity(Quantity.of(100))
            .build();
        fakeCouponStockRepository.save(stock1);
        fakeCouponStockRepository.save(stock2);
    }

    @Test
    void 쿠폰의_재고를_줄인다() {
        // given
        // when
        CouponStock deductedCouponStock = couponStockManager.deductStock(1L);
        // then
        then(deductedCouponStock.getRemainingQuantity()).isEqualTo(Quantity.of(0));
    }

    @Test
    void 동시에_100개의_쿠폰의_재고를_감소시킨다() throws InterruptedException {
        // given
        int threadCount = 100;
        ExecutorService executorService = Executors.newFixedThreadPool(32);

        CountDownLatch latch = new CountDownLatch(threadCount);
        // when
        for(int i =0; i< threadCount; i++) {
            executorService.execute(() -> {
                try {
                    couponStockManager.deductStock(2L);
                } finally {
                    latch.countDown();
                }
            });
        }

        latch.await();

        CouponStock stock = stockRepository.findByCouponId(2L).orElseThrow();
        // then
        then(stock.getRemainingQuantity().value).isEqualTo(0);
    }

}

File: core/core-application/src/test/kotlin/com/nmh/commerce/coupon/CouponStockManagerTest.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.coupon.mock.FakeCouponStockRepository
import com.nmh.commerce.domain.Quantity
import com.nmh.commerce.domain.Quantity.Companion.of
import org.assertj.core.api.BDDAssertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors

internal class CouponStockManagerTest {
    private lateinit var couponStockManager: CouponStockManager
    private lateinit var stockRepository: CouponStockRepository

    @BeforeEach
    fun setUp() {
        val fakeCouponStockRepository = FakeCouponStockRepository()
        this.stockRepository = fakeCouponStockRepository
        val couponStockManager = CouponStockManager(fakeCouponStockRepository)
        this.couponStockManager = couponStockManager
        val stock1 = CouponStock(1L, 1L, of(1), null)
        val stock2 = CouponStock(2L, 2L, of(100), null)
        fakeCouponStockRepository.save(stock1)
        fakeCouponStockRepository.save(stock2)
    }

    @Test
    fun 쿠폰의_재고를_줄인다() {
        // given
        // when
        val deductedCouponStock = couponStockManager.deductStock(1L)
        // then
        BDDAssertions.then<Quantity?>(deductedCouponStock.remainingQuantity).isEqualTo(of(0))
    }

    @Test
    @Throws(InterruptedException::class)
    fun 동시에_100개의_쿠폰의_재고를_감소시킨다() {
        // given
        val threadCount = 100
        val executorService = Executors.newFixedThreadPool(32)

        val latch = CountDownLatch(threadCount)
        // when
        for (i in 0..<threadCount) {
            executorService.execute(Runnable {
                try {
                    couponStockManager.deductStock(2L)
                } finally {
                    latch.countDown()
                }
            })
        }

        latch.await()

        val stock: CouponStock = stockRepository.findByCouponId(2L)
        // then
        BDDAssertions.then(stock.remainingQuantity.value).isEqualTo(0)
    }
}